### ISSUE TITLE
other: open TODOs

### ISSUE THREAD

**Issue by georglauterbach** [docker-mailserver/docker-mailserver#3289]
### Subject

I would like to contribute to the project

### Description

I'd like v13.0.0 to be a release that barely adds new functionality, because I think DMS is already packed with functionality. For the far future, I'd like to remove unnecessary features as much as possible (i.e. OpenDKIM, OpenDMARC, Amavis (plus codecs), Spamassassin, Postgrey, Razor, Pyzor, fts-xapian and the likes). With v13.0.0, the default will change to Rspamd. Later, with upcoming major releases after v13.0.0, we can remove functionality.

The script are already high-class, but they lack Bash's integrity and error checks - this should changes as well. Then there are many TODO tickets in the issue tracker that should be resolved.

---

- [ ] Make Rspamd the default, and change default of `1` to `0` for OpenDKIM, OpenDMARC, policyd-spf, Amavis & SpamAssassin
- [x] clean up the issue tracker
- [x] clean up stale PRs (and force-close those that have stalled for a very long time)
- [x] ~~introduce `set -u` and `set -eE` to our scripts to make them more robust~~ (EDIT: could be done, but I lack the time) 
- [x] introduce `set -o pipefail` and `shopt -s inherit_errexit` to our scripts to make them more robust
- [x] ~~deprecate `OVERRIDE_HOSTNAME` and introduce `DMS_FQDN`~~

**Comment by georglauterbach** [comment]
@polarathene could you give me a very rough (short) estimation of what it would take (also time wise) to update our CI so that we can take advantage of the Docker build cache for the ARM64 image as well?

**Comment by polarathene** [comment]
@georglauterbach it's been a while since I thought about it. I recall two options:
- Switching to the beta / experimental [GHA cache driver](https://docs.docker.com/build/cache/backends/gha/). If you look up the PRs where we introduced caching workflow or improvements, I think I provided some additional info/insight there at the time.
- Using the current [local cache driver](https://docs.docker.com/build/cache/backends/local/) with separate job and build cache key (probably prefix with `arm64` to not overlap/mix with `amd64` cache?. Then build via separate job.

While only one build cache can be exported, [multiple can be imported](https://docs.docker.com/build/cache/backends/#multiple-caches). ~~If you perform these builds locally, or in your own fork on Github you should be able to inspect the cache location (_I think the artifact can be downloaded manually via Actions cache management page_). Depending on how it exports the cache, you might be able to upload the amd64 / arm64 build cache to separate keys.~~

I think if we are sure we build the image for `amd64` / `arm64`, that earlier stage has uploaded the artifact of that platform build cache, and for publishing where it's relevant, you can just use the cache-from of both platform locations?

[`build-push-action` docs](https://github.com/docker/build-push-action#customizing) seems to support multiple `--cache-from` via `cache-from` list value. [Looks like it should be fine as multi-line yaml](https://github.com/docker/build-push-action/issues/680#issuecomment-1234458700).

You might want to subscribe to [this issue](https://github.com/actions/runner-images/issues/5631) to be notified of ARM64 runner support if it ever arrives. But hopefully the caching in CI will be good enough :+1: 

---

For current cache approach, I think you could use (_and using a platform subdir for the `path` key_):
- `cache-buildx-amd64-${{ steps.derive-image-cache-key.outputs.digest }}`
- `cache-buildx-arm64-${{ steps.derive-image-cache-key.outputs.digest }}`

https://github.com/docker-mailserver/docker-mailserver/blob/bba72daedf4ecf03a80eea571160371191237863/.github/workflows/generic_publish.yml#L51-L52

Then:

https://github.com/docker-mailserver/docker-mailserver/blob/bba72daedf4ecf03a80eea571160371191237863/.github/workflows/generic_publish.yml#L84

```yaml
cache-from: |
  type=local,src=/tmp/.buildx-cache/amd64
  type=local,src=/tmp/.buildx-cache/arm64
```

Just note that other workflows are using [the derived key](https://github.com/docker-mailserver/docker-mailserver/blob/bba72daedf4ecf03a80eea571160371191237863/.github/workflows/generic_build.yml#L58) from the build workflow to pull in a cache too. You need to make sure you're properly handling the `amd64`/`arm64` separate build caches for build workflow in other workflows such as the test PR one.

The build workflow should be fine if it's not doing mixed platform build, then it only needs to update the cache key and `path` can be left as-is. Other jobs that use `generic-build.yml` can then just have the `build-push-action` step load in the caches? I think only the publish workflow needs both AMD64 and ARM64 for `build-push-action` to have been built prior for cache use, while the testing workflow only needs `generic-build.yml` cache key to be adjusted to be platform specific (_could use `${{ inputs.platforms }}` if the input was always only `amd64` or `arm64` and not both?_).

**Comment by georglauterbach** [comment]
Thank you very much! ❤️ I will, after much of the above work is done, consider the GHA build cache to finally bring down overall PR test time.

**Comment by casperklein** [comment]
tl;dr;

Providing alternatives 👍 
Changing DMS defaults 👍 
Removing anything soon: 👎

---

> Later, with upcoming major releases after v13.0.0, we can remove functionality.

Just for the record. We should do this IMO very conservative. I am fine with providing alternatives like fetchmail --> getmail. 
The same goes for rspamd which could supersede more than one service.

My reasoning is, at the moment, there are a lot of setups (mine included), that just works. For example, I've been using fetchmail for some time and didn't run in any issue ever. So from a rational point, there is absolutely no reason, why I should switch to getmail (as long as fetchmail is officially supported and receives security updates if required). Switching could only introduce new issues, that haven't been any before. Never change a running system 😎 
Rspamd is even more critical to switch to, because it not only replaces SA, but also offers additional things like SPF/DKIM verification. It reminds me a bit of systemd..
Personally I really like the approach, 1 service for 1 thing. Instead of 1 service (+ modules) trying to do many things. I like it more isolated.
At the end, I may migrate to rspamd, that is not my point. But it will introduce additional work for me, with (at first glance) no benefit and potential to encounter issues, that I haven't before. My mailserver is critical to me and must work. Migration must be carefully planed, tested and monitored. At the moment, I don't have much time / willing to spend it for a migration.

**Comment by georglauterbach** [comment]
Agreed 👍🏼 the whole thing was more about switching. Before we actually remove functionality, there are going to be a lot of releases in between I suspect. I wouldn't change a running system either.

**Comment by polarathene** [comment]
@georglauterbach did you want this issue closed? Or should it be opened and drop the version specific target?

**Comment by georglauterbach** [comment]
> @georglauterbach did you want this issue closed? Or should it be opened and drop the version specific target?

Good catch. But I'm thinking about opening a new one and transferring the leftover TODOs to have a new clean state.

**Comment by polarathene** [comment]
> before we integrate new features like Vector, I'd like to see this issue being cleared.

Yes absolutely I agree 👍 

I only raised that issue as it was something I had been meaning to better document as a TODO that I could delegate. I don't expect it to be a priority any time soon.

> I'd be very nice of we can also take care of the issues that are not yet completed mentioned in your comments :)

I'll see what I can do, current priorities are:

- Finish LDAP rework + docs.
- Find my local repo with Dovecot test PR and get that over the finish line (_this fixes the Dovecot quota skipped tests IIRC_).
- Get the encryption PR pushed (_I've found this from an old VM_).
- `chown -R` with `/var/mail` issue to be addressed. I've documented this with a TODO recently, and I need that encryption PR sorted before this as I recall it being affected by this logic too. Ideally after the Dovecot PR is handled, I can quickly convert the remaining test suites and look into how much effort it would be to PR the next rework of the test suite to leverage `compose.yaml` features for matrix tests.
- Ideally address the `SPOOF_PROTECTION` concerns issue. I had looked into that earlier but had to detour. I identified the issue with LDAP test, which requires a breaking change when using `SASLAUTHD` to use `-r` for the `ldap` mechanism, like `rimap` mechanism does. The logic for Postfix config also needs some changes, IIRC since introducing the `dms_` config var into `main.cf` for submission(s) ports to use, the protection was no longer applicable to port 25 which had some relevance 😅 
- After addressing the `SPOOF_PROTECTION` issue, following up with docs related to that. At this point, we may want to consider removing the `setup alias` guard, as we have an open issue where it seemed valid to support a scenario (_deliver to local inbox of same address and forward via alias_), at least I was not aware of any other way to support it (_except with ldap and a `postfix-book.schema`_).

If there's any other pressing TODO items beyond that, let me know. I'd like to tackle the `DMS_FQDN` one, along with the follow-up refactor of TLS startup scripts support (_plus docs page refactoring_). Probably delaying to v14 with current workload however. I've made sure to raise issues for these at least in the event I become unavailable.

The LDAP rework will close several long-standing issues, along with the `chown -R` issue being resolved. It seems best to address all of these LDAP related concerns for v13 since they're breaking changes. It should also become possible to better integrate with `check-for-changes.sh` for certificate updates.

**Comment by polarathene** [comment]
@zdenbe I don't know what you're doing with this DMS fork, but you recently contributed a PR there with no context that seems to be commits since `10.5` to `12.1`: https://github.com/evymo/docker-mailserver/pull/1

That created a reference to many issues / PRs due to your [merge commit message](https://github.com/evymo/docker-mailserver/commit/8c1aeb0f95b4c921e1a583a052f4d8fd48eada56) using our merge commits and their messages. Please don't do that.

**Comment by georglauterbach** [comment]
Perfect! Let me know if I can assist you somewhere (preferably not LDAP 😆)

**Comment by polarathene** [comment]
> Let me know if I can assist you somewhere

If you've got time and any of these interest you, they'd probably be good to tackle:

- https://github.com/docker-mailserver/docker-mailserver/issues/3558
- https://github.com/docker-mailserver/docker-mailserver/issues/3280
- https://github.com/docker-mailserver/docker-mailserver/issues/3478

I didn't get around to checking, but for `PERMIT_DOCKER` I noticed rspamd had some IP subnets being configured, might be related to what `PERMIT_DOCKER` is meant to control? (_there's similar for Dovecot too I think_)

The sieve and amavis issues are probably a good fit for you. The `PERMIT_DOCKER` one is probably a hassle to tackle, but it'd be good to have your input regarding if rspamd network trust should be managed by it like OpenDKIM / OpenDMARC is.

**Comment by zdenbe** [comment]
I am sorry. I dont know how its happened… My fault. Nothing to contribute to origin. I tried just update fork… GLZdeněk Bělka13. 10. 2023 v 3:05, Brennan Kinney ***@***.***>:﻿
@zdenbe I don't know what you're doing with this DMS fork, but you recently contributed a PR there with no context that seems to be commits since 10.5 to 12.1: evymo#1
That created a reference to many issues / PRs due to your merge commit message using our merge commits and their messages. Please don't do that.

**Comment by georglauterbach** [comment]
@polarathene I'm trying to close some old, stale issues of mine, I also came across this one. I will implement the changes for then defaults for Rspamd; could you take a look at the open checkboxes from [this comment](https://github.com/docker-mailserver/docker-mailserver/issues/3289#issuecomment-1521333815) please and tell me what we should do about them? I would like to close this issue at some point; otherwise I'll go insane with issues.

**Comment by polarathene** [comment]
> I will implement the changes for then defaults for Rspamd

I'm fine with switching to rspamd for v15, but I'd like a v14.1 release first.

I have been spending time over the past week with a docs rewrite, I should have the PR for that up by the weekend. I really want the LDAP PR off my mind 😅 

> could you take a look at the open checkboxes from [this comment](https://github.com/docker-mailserver/docker-mailserver/issues/3289#issuecomment-1521333815)

`SPOOF_PROTECTION` related tasks I'll tackle after the LDAP PR is finally out of the way. Necessary for the related PR being contributed for me to review and wrap all that up (_nobody else was willing to tackle it properly thus far_).

> please and tell me what we should do about them?

That leaves two docs related tasks, I can't commit to anything extra right now, especially docs which are the slowest tasks for me to tackle. I could raise them as separate issues for discoverability perhaps, but otherwise I expect they'll go stale and closed if I don't address them.

---
## Off-topic

I've been thinking about what my cut-off point as a maintainer looks like, it's not realistic for me to resolve all the tasks I'm aware of (_as much as I'd like to_). It's getting difficult to justify the value of the efforts when there is alternatives like Stalwart and MailCow available.

> I would like to close this issue at some point; otherwise I'll go insane with issues.

I can kind of relate, although the number of open issues don't bother me so much. Closing them doesn't do much good, but given the lack of new contributors (_even when I info dump context to assist_), the issues may not get resolved, or some lower quality fix is introduced and approved like before I was a maintainer which caused some of the tech debt problems we have today in DMS😅 

My existing commitments and especially unfinished/blocked contributions are what stress me out. The issues and time I spent looking into understanding and documenting them along with how to proceed, are also regularly on my mind about when I'll resolve them with a PR. I don't want to let users down, nor see my time invested thus far wasted away, but lately the awareness of how much time it'll take me to tackle it all when I should be spending that time on myself (_instead of volunteering for projects I don't even use_ 😂 ), I realize I need to get comfortable letting it go.

So if you want to close issues, go for it. Maintaining projects can take a toll, if that helps you avoid burnout by focusing only on the critical issues and deferring community to step up for tackling the others, I get that 👍

**Comment by georglauterbach** [comment]
I'm absolutely with you on this one. As you can see, I haven't had time over the course of the last weeks to do anything meaningful for this project, and I think that's fine. We're all spending our free time here, and no one is obligated to contribute. The reason I came back to this was purely for cleaning up my issues.

A word of advice, that I should pay attention to, too: There are many things that can be done to improve this project even further, but focusing on the most important issues takes more than enough time for all of us. I think you provide a commitment to this project that is simply too big, and I'd recommend you to take it easy with answering issues and discussions :) It's awesome for the people that seek help, but they can in most cases get this information from somewhere else too. Even if not, your time spent may be too big; hence, I can only recommend doing a little less - that's already more than enough. This way, when you come back after a bit of time, and if you're still interested in doing something, you can do so with a good feeling. Otherwise, you may regret spending too much time on DMS.

IMO: DMS provides people all over the world with an awesome experience already; and simply providing easy maintenance work at the moment is more than fine! If someone want a feature, they need to go ahead and work on it themselves at the moment - we cannot solve all the problems that others may have or implement features others may want.

Maintainer fatigue is a real problem, and also prevalent, e.g., in the Linux kernel community. I'd like to see every maintainer stay fit and healthy, and we need to avoid at all costs that core maintainers suffer by maintaining this project. I think DMS is in a comfortable state where many things are done right, not least due to @casperklein, your and my contributions.

---

I, myself, will of course continue to maintain DMS, but with a lower time commitment than in the last years. This way, I stay healthy, and if an important issue comes around, I can tackle it without having already spent too much time.

---

In the light of things, I will close this issue and **consider all items here as "will not be implemented". This should free us from thinking of them as work that we still **need** to do. If we ever decide to tackle them, we can do so; if not, that's absolutely fine.
